src/caffe/layers/weighted_euclidean_loss_layer.cpp

```diff
@@ -0,0 +1,67 @@
+#include <vector>
+
+#include "caffe/layers/weighted_euclidean_loss_layer.hpp"
+#include "caffe/util/math_functions.hpp"
+
+namespace caffe {
+
+template <typename Dtype>
+void WeightedEuclideanLossLayer<Dtype>::Reshape(
+  const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  LossLayer<Dtype>::Reshape(bottom, top);
+  CHECK_EQ(bottom[0]->count(1), bottom[1]->count(1))
+      << "Inputs must have the same dimension.";
+  diff_.ReshapeLike(*bottom[0]);
+}
+
+template <typename Dtype>
+void WeightedEuclideanLossLayer<Dtype>::Forward_cpu(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top) {
+  int count = bottom[0]->count();
+  caffe_sub(
+      count,
+      bottom[0]->cpu_data(),
+      bottom[1]->cpu_data(),
+      diff_.mutable_cpu_data());
+
+  Dtype wdot(0.0);
+  for (int i = 0; i < count; ++i) {
```

Comment:
You can compute the loss in a slightly different way (assuming your weights are non-negative):
1. compute diff <- (x1-x2)
2. scale the diff by **sqrt** of the weights: sdiff <- sqrt(x).*diff
3. compute the dot product of <sdiff,sdiff>

These three steps can be easily carried out using off-the-shelf math functions that can be found in `math_functions.cpp`.
Once your forward/backward functions are implemented using caffe "math functions" it should be fairly easy to implement a GPU version of the layer - this can be really important for successful usage of the layer.